R/particle_filter_data.R

```diff
@@ -69,6 +97,95 @@ particle_filter_data <- function(data, time, rate, initial_time = NULL) {
   }
 
   rate <- assert_scalar_positive_integer(rate)
+  t <- clean_pf_times(t, initial_time, data)
+
+  ret <- data.frame(time_start = t$time_start,
+                    time_end = t$time_end,
+                    step_start = t$time_start * rate,
+                    step_end = t$time_end * rate,
+                    data[names(data) != time])
+  names(ret)[1:2] <- paste0(time, c("_start", "_end"))
+  attr(ret, "rate") <- rate
+  attr(ret, "time") <- time
+  class(ret) <- c("particle_filter_data_single", "particle_filter_data",
+                  "data.frame")
+
+  ret
+}
+
+particle_filter_data_nested <- function(data, time, rate, initial_time,
+                                        population, allow_unequal_times) {
+  # catch invalid user call
+  if (is.null(population)) {
+    stop("'population' must be non-NULL")
+  }
+
+  if (!is.factor(data[[population]])) {
+    stop(sprintf("Column '%s' must be a factor", population))
+  }
+
+  groups <- unique(data[[population]])
+
+  assert_logical(allow_unequal_times)
+
+  split_data <- split(data[names(data) != population], data[[population]])
+  vars <- names(data)[!(names(data) %in% c(time, population))]
+
+  nr <- viapply(split_data, nrow)
+  if (length(unique(nr)) > 1) {
+    if (allow_unequal_times) {
+      ## assumed quicker handling one matrix then re-splitting than lapply
+      times <- as.integer(sort(unique(unlist(lapply(split_data, "[[", time)))))
+      split_data <- lapply(split_data, function(x) {
+        df <- data.frame(matrix(ncol = ncol(x) - 1, nrow = length(times)))
+        colnames(df) <- colnames(x)[-ncol(x)]
+        df[[time]] <- times
+        df[match(x[[time]], times), vars] <- as.matrix(x[, vars])
+        df
+      })
+    } else {
+      stop("Unequal time between populations")
+    }
+  }
+
+  rate <- assert_scalar_positive_integer(rate)
+  ## times now equal so select first arbitrarily
+  t <- clean_pf_times(split_data[[1]][[time]], initial_time, data)
+
+  data <- do.call(rbind, split_data)
+
+  ret <- data.frame(time_start = t$time_start,
+                    time_end = t$time_end,
+                    step_start = t$time_start * rate,
+                    step_end = t$time_end * rate,
+                    population = rep(groups, each = length(t$time_start)),
+                    data[vars])
```

Comment:
stringsAsFactors = FALSE here to prevent possible coersion of strings in `data[vars]` in R < 4